File: examples/atb_models/tests/fuzztest/core/test_word_embedding.cpp
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
 * MindIE is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

#include <gtest/gtest.h>
#include <atb/atb_infer.h>
#include "atb_speed/log.h"
#include "operations/fusion/embedding/word_embedding.h"
#include "../utils/fuzz_utils.h"
#include "secodeFuzz.h"

namespace atb_speed {
TEST(WordEmbeddingDTFuzz, WordEmbedding)
{
    std::srand(time(NULL));
    ATB_SPEED_LOG_DEBUG("begin====================");
    std::string fuzzName = "WordEmbeddingDTFuzzWordEmbedding";

    DT_FUZZ_START(0, 10000, const_cast<char*>(fuzzName.c_str()), 0) {
        uint32_t fuzzIndex = 0;

        int worldSize = *(int *) DT_SetGetNumberRange(&g_Element[fuzzIndex++], 1, 1, 8);
        std::vector<std::string> backendEnumTable = {"lccl", "hccl"};

        atb_speed::common::WordEmbeddingParam param;
        param.unpadInputs = FuzzUtil::GetRandomBool(fuzzIndex);
        param.tensorParallelInfo.rank = *(int *) DT_SetGetNumberRange(&g_Element[fuzzIndex++], 0, 0, worldSize);
        param.tensorParallelInfo.worldSize = worldSize;
        param.tensorParallelInfo.backend = backendEnumTable[*(int *) DT_SetGetNumberRange( \
            &g_Element[fuzzIndex++], 0, 0, 1)];

        atb::Node wordEmbeddingNode;
        WordEmbedding(param, &wordEmbeddingNode.operation);
    }
    DT_FUZZ_END()
    SUCCEED();
}
}

File: examples/atb_models/tests/pythontest/atb_llm/layers/embedding/test_word_embedding.py
#!/usr/bin/env python
# coding=utf-8
# Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
# MindIE is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#          http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.

import unittest
from unittest.mock import MagicMock, patch

import torch

from atb_llm.layers.embedding.word_embedding import ParallelEmbedding
from atb_llm.utils.loader.safetensor_file_loader import SafetensorFileLoader
from atb_llm.models.base.config import BaseConfig
from atb_llm.nn.tensor import Tensor


class TestWordEmbedding(unittest.TestCase):
    def setUp(self):
        self.config = BaseConfig(torch_dtype=torch.float16)

    @patch("atb_llm.layers.embedding.word_embedding")
    def test_word_embedding(self, mock_ops):
        weight_tool_cls = MagicMock(spec=SafetensorFileLoader)
        mock_weight_tool_obj = weight_tool_cls()
        embedding_tensor = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float16)
        mock_weight_tool_obj.get_tensor = MagicMock(return_value=embedding_tensor)
        mapping = MagicMock()
        module = ParallelEmbedding(self.config, mock_weight_tool_obj, "word_embedding", mapping)
        input_ids = Tensor("input_ids")
        golden_output = Tensor()
        mock_ops.gather.return_value = golden_output
        out = module.forward(input_ids)
        self.assertEqual(golden_output, out)
    
    def test_parallel_embedding(self):
        weight_tool_cls = MagicMock(spec=SafetensorFileLoader)
        mock_weight_tool_obj = weight_tool_cls()
        embedding_tensor = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float16)
        mock_weight_tool_obj.get_sharded = MagicMock(return_value=embedding_tensor)
        mapping = MagicMock()
        _ = ParallelEmbedding(self.config, mock_weight_tool_obj, "word_embedding",
                                   mapping, parallel_embedding=True)


if __name__ == '__main__':
    unittest.main()